allocsim: add a duration and final score

BramGruneir opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12179)
<!-- Reviewable:end -->

BramGruneir committed:
allocsim: add a duration and final score

petermattis commented on pkg/cmd/allocsim/main.go:
```diff
@@ -29,17 +30,18 @@ import (
 
 var workers = flag.Int("w", 1, "number of workers; the i'th worker talks to node i%numNodes")
 var numNodes = flag.Int("n", 4, "number of nodes")
+var timeout = flag.Duration("timeout", math.MaxInt64, "how long to run allocsim for")
```
In other tools we call this `duration`. For example, `block_writer -duration`.

petermattis reviewed (approved):
LGTM

BramGruneir commented:
PTAL
Added a very basic summary/scorecard at the end of the run.  

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/cmd/allocsim/main.go, line 44 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYUJGfMKh9A0aybafBM-r1-44:-KYV1R39je5M3sHCJY8p:b-uzawu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f571e95b3403d41ab5992cfbb04d1847f096682e/pkg/cmd/allocsim/main.go#L44)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

In other tools we call this `duration`. For example, `block_writer -duration`.
</blockquote></details>

Changed to duration.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-:-KYV1a3goBqwsH0TEFsT:b-acqgbc)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/allocsim/main.go, line 44 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYUJGfMKh9A0aybafBM-r1-44:-KYV39oYPczW3amUB6Rt:b-95gqwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f571e95b3403d41ab5992cfbb04d1847f096682e/pkg/cmd/allocsim/main.go#L44)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Changed to duration.
</blockquote></details>

`s/timeout/duration/g`

---

*[pkg/cmd/allocsim/main.go, line 46 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV4Td5eEryoqye-aaV:-KYV4Td5eEryoqye-aaW:b-1n507i) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L46)):*
> ```Go
> var timeout = flag.Duration("duration", math.MaxInt64, "how long to run the simulation for")
> var blockSize = flag.Int("b", 1000, "block size")
> var rebalanceThreshold = flag.Float64("rebalanceThreshold", 0.05, "acceptable range for leases and replicas to deviate from the mean")
> ```

The rebalancer allows replicas and leases to be up to 5% from the mean. So the total distance is 10%.

---

*[pkg/cmd/allocsim/main.go, line 246 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV4ABqiSbj-nN6cS3M:-KYV4ABqiSbj-nN6cS3N:b-4fdm96) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L246)):*
> ```Go
> 
> func getStats(counts []int) (min, max int, mean, highestDiff float64) {
> 	min = math.MaxInt16
> ```

`s/MaxInt16/MaxInt32/g`

---

*[pkg/cmd/allocsim/main.go, line 277 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV3f8hVBt99pQX4AaU:-KYV3f8hVBt99pQX4AaV:bagrikh) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L277)):*
> ```Go
> 	minReplicas, maxReplicas, meanReplicas, replicasDiff := getStats(a.ranges.replicas)
> 	fmt.Printf("Replicas:\n")
> 	fmt.Printf("\tMinimum: %d\n", minReplicas)
> ```

Would be nice of the columns lined up here. Use `tabwriter`.

---

*[pkg/cmd/allocsim/main.go, line 280 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV4FvmnXLIDYuRel4o:-KYV4FvnaC4h6kpVjtMI:b-msm3sy) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L280)):*
> ```Go
> 	fmt.Printf("\tMaximum: %d\n", maxReplicas)
> 	fmt.Printf("\tMean: %.1f\n", meanReplicas)
> 	fmt.Printf("\tHighest Percent Diff: %.1f%%\n", replicasDiff*100)
> ```

`s/Highest/Maximum

---

*[pkg/cmd/allocsim/main.go, line 281 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV3aP854HqUm5QkyUf:-KYV3aP854HqUm5QkyUg:b-q3kzww) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L281)):*
> ```Go
> 	fmt.Printf("\tMean: %.1f\n", meanReplicas)
> 	fmt.Printf("\tHighest Percent Diff: %.1f%%\n", replicasDiff*100)
> 	fmt.Printf("\tScore: %t\n", replicasDiff < *rebalanceThreshold)
> ```

Score is a boolean? I'm not sure how useful this is. You can visually inspect the previous line to see if the max diff is too far.

---

*[pkg/cmd/allocsim/main.go, line 289 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV4cH38vM17L4_oy6Z:-KYV4cH38vM17L4_oy6_:b2gvopo) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L289)):*
> ```Go
> 	fmt.Printf("\tMean: %.1f\n", meanLeases)
> 	fmt.Printf("\tHighest Percent Diff: %.1f%%\n", leasesDiff*100)
> 	fmt.Printf("\tScore: %t\n", leasesDiff < *rebalanceThreshold)
> ```

Rather than this score card, perhaps the final report should show the per-node replicas/leases as percentages and another line as their distance from the mean.

---

*[pkg/cmd/allocsim/main.go, line 307 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV3mG0iovjlwd0kCVC:-KYV3mG0iovjlwd0kCVD:b-txoabm) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L307)):*
> ```Go
> 
> 	c.Start("allocsim", *workers, flag.Args(),
> 		[]string{"COCKROACH_METRICS_SAMPLE_INTERVAL=2s"})
> ```

Should remove the env var now that we have reactive gossip.

---

*[pkg/cmd/allocsim/main.go, line 308 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV3v_6-li8mJrRE_sg:-KYV3v_6-li8mJrRE_sh:b-4q1h1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L308)):*
> ```Go
> 	c.Start("allocsim", *workers, flag.Args(),
> 		[]string{"COCKROACH_METRICS_SAMPLE_INTERVAL=2s"})
> 	c.UpdateZoneConfig(1, 1<<20)
> ```

Why did you move this? Now there is a window of time in which you can `ctrl-C` the process and it would be cleaned up propertly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Updated, sample output
```
Total Ranges: 180
+-------------+------+------+------+------+
|  Replicas   |  1   |  1   |  1   |  1   |
+-------------+------+------+------+------+
|     % total | 23.9 | 25.0 | 25.0 | 26.1 |
| % from mean |  4.4 |  0.0 |  0.0 |  4.4 |
+-------------+------+------+------+------+
        Minimum 43
        Maximum 47
           Mean 45.0
   Maximum Diff 4.4%
+-------------+------+------+------+------+
|   Leases    |  1   |  1   |  1   |  1   |
+-------------+------+------+------+------+
|     % total | 33.3 | 19.3 | 17.5 | 29.8 |
| % from mean | 33.3 | 22.8 | 29.8 | 19.3 |
+-------------+------+------+------+------+
        Minimum 10
        Maximum 19
           Mean 14.2
   Maximum Diff 33.3%
```

---

Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/cmd/allocsim/main.go, line 44 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYUJGfMKh9A0aybafBM-r1-44:-KYV5gKlWVwpff14IF4_:bgcj5vu) ([raw file](https://github.com/cockroachdb/cockroach/blob/f571e95b3403d41ab5992cfbb04d1847f096682e/pkg/cmd/allocsim/main.go#L44)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/timeout/duration/g`
</blockquote></details>

ha, missed the variable.  Fixed.

---

*[pkg/cmd/allocsim/main.go, line 46 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV4Td5eEryoqye-aaV:-KYV6IZUwV8z7a2pkiPv:bg9ucug) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L46)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The rebalancer allows replicas and leases to be up to 5% from the mean. So the total distance is 10%.
</blockquote></details>

removed anyway

---

*[pkg/cmd/allocsim/main.go, line 246 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV4ABqiSbj-nN6cS3M:-KYV6OALqKlWcmY_HSEv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L246)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/MaxInt16/MaxInt32/g`
</blockquote></details>

Done.

---

*[pkg/cmd/allocsim/main.go, line 277 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV3f8hVBt99pQX4AaU:-KYVEHJj5C0cX2inrfmr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L277)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Would be nice of the columns lined up here. Use `tabwriter`.
</blockquote></details>

Done.

---

*[pkg/cmd/allocsim/main.go, line 280 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV4FvmnXLIDYuRel4o:-KYV6VfY35_crYn1qrWr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L280)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/Highest/Maximum
</blockquote></details>

Done.

---

*[pkg/cmd/allocsim/main.go, line 281 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV3aP854HqUm5QkyUf:-KYV6XVRrjmc4zdM4U_T:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L281)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Score is a boolean? I'm not sure how useful this is. You can visually inspect the previous line to see if the max diff is too far.
</blockquote></details>

removed

---

*[pkg/cmd/allocsim/main.go, line 289 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV4cH38vM17L4_oy6Z:-KYVDUvB-isSh00Lb7Ns:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L289)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Rather than this score card, perhaps the final report should show the per-node replicas/leases as percentages and another line as their distance from the mean.
</blockquote></details>

Done.

---

*[pkg/cmd/allocsim/main.go, line 307 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV3mG0iovjlwd0kCVC:-KYVDyycJ5sa6OBre8Hw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L307)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Should remove the env var now that we have reactive gossip.
</blockquote></details>

Done.

---

*[pkg/cmd/allocsim/main.go, line 308 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV3v_6-li8mJrRE_sg:-KYVE0-6IfTs08cvNM3f:bbr12fg) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L308)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Why did you move this? Now there is a window of time in which you can `ctrl-C` the process and it would be cleaned up propertly.
</blockquote></details>

needed allocsim `a` to exist to call finalStatus on it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-:-KYVEIXdxBRYjydl9EVY:bq2u8vs)*
<!-- Sent from Reviewable.io -->

petermattis commented:
It would be nice of the final output lined up with the normal output. I'm not seeing a strong reason for a completely different style.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/allocsim/main.go, line 308 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV3v_6-li8mJrRE_sg:-KYVM7NnmZdbV0sBNldY:b-vr79zz) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L308)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

needed allocsim `a` to exist to call finalStatus on it.
</blockquote></details>

Move the calls to `c.Start` and `c.UpdateZoneConfig` back down below. The call to `newAllocSim` can remain.

---

*[pkg/cmd/allocsim/main.go, line 263 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYVLnlsAlyBgNBwgvzY:-KYVLnltUsMDrQgyBhZX:b-miy2gx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f27b83a94edc00405556e8a3601384c78d79cf9/pkg/cmd/allocsim/main.go#L263)):*
> ```Go
> 	percentsRow[0] = "% total"
> 	diffsRow := make([]string, len(counts)+1)
> 	diffsRow[0] = "% from mean"
> ```

`s/% from mean/|%(count - avg)|/g`?

---

*[pkg/cmd/allocsim/main.go, line 291 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYVK55w_d2bFW-PIdUA:-KYVK55w_d2bFW-PIdUB:b-5qd0x5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f27b83a94edc00405556e8a3601384c78d79cf9/pkg/cmd/allocsim/main.go#L291)):*
> ```Go
> 	fmt.Fprintf(w, "Maximum\t %d\n", max)
> 	fmt.Fprintf(w, "Mean\t %.1f\n", mean)
> 	fmt.Fprintf(w, "Maximum Diff\t %.1f%%\n", maxDiff)
> ```

These 4 lines of stats aren't terribly useful. The `Maximum Diff` is easily determined from the table and the other values are absolute and not percentages. The `Minimum` and `Maximum` values can be easily determined from the last row of normal output.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-:-KYVMRsGJYuCDXGiD9gG:b-lskhcz)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Ok, rewrote all the output to use the tablewriter, it's not ideal, since it won't let me pre-define each column size, but it's a lot more uniform and gives us the nice boarders. 

I might try to experiment producing the same results with the tabwriter, might make the code a little bit cleaner.  Not sure.

```
+----------+----------+----------+----------+-----------+-----------+-----------+-----------+
  elapsed  | ops/sec  |  errors  | replicas |     0     |     1     |     2     |     3
+----------+----------+----------+----------+-----------+-----------+-----------+-----------+
     1s    |  1871.6  |    0     |    18    |    6/0    |    4/0    |    4/0    |    4/0
     2s    |  1935.8  |    0     |    24    |    8/0    |    6/0    |    4/0    |    6/0
     3s    |  1805.8  |    0     |    30    |   10/0    |    8/0    |    4/0    |    8/0
     4s    |  1836.8  |    0     |    45    |   13/0    |   11/0    |   10/0    |   11/0
     5s    |  1866.0  |    0     |    42    |   10/0    |   11/0    |   10/0    |   11/0
     6s    |  1956.8  |    0     |    66    |   17/0    |   17/0    |   16/0    |   16/0
     7s    |  1962.5  |    0     |    66    |   17/0    |   17/0    |   16/0    |   16/0
     8s    |  1905.7  |    0     |    66    |   17/0    |   17/0    |   16/0    |   16/0
     9s    |  1751.8  |    0     |    69    |   18/0    |   18/0    |   16/0    |   17/0
    10s    |  1835.5  |    0     |   108    |   27/0    |   27/0    |   28/0    |   26/0
    11s    |  1922.6  |    0     |   114    |   29/14   |   28/8    |   29/3    |   28/9
    12s    |  1915.7  |    0     |   114    |   29/14   |   28/8    |   29/3    |   28/9
    13s    |  1963.8  |    0     |   114    |   29/14   |   28/8    |   29/3    |   28/9
    14s    |  1961.8  |    0     |   114    |   29/14   |   28/8    |   29/3    |   28/9
    15s    |  1892.1  |    0     |   114    |   29/14   |   28/8    |   29/3    |   28/9
    16s    |  1920.3  |    0     |   114    |   29/14   |   28/8    |   29/3    |   28/9
    17s    |  1781.9  |    0     |   123    |   32/14   |   31/8    |   30/3    |   30/9
    18s    |  1808.7  |    0     |   162    |   38/14   |   42/8    |   42/3    |   40/9
    19s    |  1840.5  |    0     |   189    |   47/14   |   47/8    |   48/3    |   47/9
    20s    |  1895.6  |    0     |   210    |   53/24   |   52/15   |   53/11   |   52/16
+----------+----------+----------+----------+-----------+-----------+-----------+-----------+
  elapsed  | ops/sec  |  errors  | replicas |     0     |     1     |     2     |     3
+----------+----------+----------+----------+-----------+-----------+-----------+-----------+
    21s    |  1786.4  |    0     |   210    |   53/24   |   52/15   |   53/11   |   52/16
    22s    |  1924.1  |    0     |   210    |   53/24   |   52/15   |   53/11   |   52/16
    23s    |  1790.4  |    0     |   210    |   53/24   |   52/15   |   53/11   |   52/16
    24s    |  1932.2  |    0     |   210    |   53/24   |   52/15   |   53/11   |   52/16
    25s    |  1873.6  |    0     |   210    |   53/24   |   52/15   |   53/11   |   52/16
    26s    |  1932.6  |    0     |   210    |   53/24   |   52/15   |   53/11   |   52/16
    27s    |  1862.6  |    0     |   210    |   53/24   |   52/15   |   53/11   |   52/16
    28s    |  1863.1  |    0     |   210    |   53/24   |   52/15   |   53/11   |   52/16
    29s    |  1805.6  |    0     |   210    |   53/24   |   52/15   |   53/11   |   52/16
+----------+----------+----------+----------+-----------+-----------+-----------+-----------+
   stats   |          |          |          |     0     |     1     |     2     |     3
+----------+----------+----------+----------+-----------+-----------+-----------+-----------+
           | replicas |          | total %  |   25.2%   |   24.8%   |   25.2%   |   24.8%
           |   210    |          |  diff %  |   0.2%    |   0.2%    |   0.2%    |   0.2%
           |  leases  |          | total %  |   32.3%   |   24.6%   |   18.5%   |   24.6%
           |    65    |          |  diff %  |   7.3%    |   0.4%    |   6.5%    |   0.4%
```

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/cmd/allocsim/main.go, line 308 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV3v_6-li8mJrRE_sg:-KYZai1QYqAQ3AsI2b1Z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L308)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Move the calls to `c.Start` and `c.UpdateZoneConfig` back down below. The call to `newAllocSim` can remain.
</blockquote></details>

Done.

---

*[pkg/cmd/allocsim/main.go, line 263 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYVLnlsAlyBgNBwgvzY:-KYZakC_5Xix7sR5IinB:bo0suqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f27b83a94edc00405556e8a3601384c78d79cf9/pkg/cmd/allocsim/main.go#L263)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/% from mean/|%(count - avg)|/g`?
</blockquote></details>

that was some funky math, all fixed.

---

*[pkg/cmd/allocsim/main.go, line 291 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYVK55w_d2bFW-PIdUA:-KYWrgSYs8eBQ-5uGuaV:bnjzsy5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f27b83a94edc00405556e8a3601384c78d79cf9/pkg/cmd/allocsim/main.go#L291)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

These 4 lines of stats aren't terribly useful. The `Maximum Diff` is easily determined from the table and the other values are absolute and not percentages. The `Minimum` and `Maximum` values can be easily determined from the last row of normal output.
</blockquote></details>

I'll just remove them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-:-KYZaop5UsG4VzBBHUdo:bk4b63i)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/cmd/allocsim/main.go, line 345 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYZeovitd4_Z6-1EwvJ:-KYZeovitd4_Z6-1EwvK:b-eggfy8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f455a177f0eb414fcea09fadc01919fc7db5277/pkg/cmd/allocsim/main.go#L345)):*
> ```Go
> 		a.finalStatus()
> 		c.Close()
> 		os.Exit(1)
> ```

is exiting (1) appropriate after the duration expires?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> Ok, rewrote all the output to use the tablewriter, it's not ideal, since it won't let me pre-define each column size, but it's a lot more uniform and gives us the nice boarders.

It was not my intention that you change the existing format, but that you make the final stats match what was already present. Re-reading what I wrote I see how you could have interpreted it differently. The borders aren't desirable when horizontal space is at a premium which it is as you move to higher number of node counts.

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/cmd/allocsim/main.go, line 308 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV3v_6-li8mJrRE_sg:-KYZflnqe-NeaRuuKHhO:bihh87l) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L308)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

This doesn't look done. This looks exactly the same.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-:-KYZf4nJQFcoEOnpkbf6:b-288n1t)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
I was just trying to figure out how to make it all a lot cleaner.  I'll remove all the boarders and spacers and try replacing it all with a tabwriter instead.

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/allocsim/main.go, line 308 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV3v_6-li8mJrRE_sg:-KYZh_xfUyXzZhFRLvZ8:b-gmwn1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L308)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This doesn't look done. This looks exactly the same.
</blockquote></details>

Hmmm... I'm sure I moved it down.  Must not have saved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-:-KYZhribgMuwzOadvi5h:b-24vd4s)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> I was just trying to figure out how to make it all a lot cleaner.  I'll remove all the boarders and spacers and try replacing it all with a tabwriter instead.

I don't think you can use tabwriter here as it buffers output. My recommendation would be to leave the existing output code alone and adapt the final stats to match that style. For example:

```

___stats____________________________________1_________2_________3_________4_________5_________6
 replicas (total%/diff%)                 16/0      16/0      16/0      16/0      16/0      16/0
 leases (total%/diff%)                   16/0      16/0      16/0      16/0      16/0      16/0
```

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-:-KYZjmeKTAjpHOyMFRtE:be63xxw)*
<!-- Sent from Reviewable.io -->

tamird commented:
tabwriter does buffer output, but it is possible to flush it manually. just sayin', since i do like the new output better =/

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-:-KY_BGHmxvgbPNvF-caI:b7cnqat)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> tabwriter does buffer output, but it is possible to flush it manually.

`tabwriter.Writer.Flush` terminates all of the columns causing the next call to write to start a new set of columns (with potentially different widths). I'm sure you could get this to produce the same output as is currently present, but why go through the effort?

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-:-KY_EPXEcyhcEQMhll__:b-2q1iua)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
I've removed the tablewriter and reformatted the output to match the original style but I did make a function so the headers and all lines can use it.
And the stats now match what you requested.

```
_elapsed__ops/sec___errors_replicas_________0_________1_________2_________3
      1s   1877.8        0       18       6/0       4/0       4/0       4/0
      2s   2000.7        0       24       8/0       6/0       6/0       4/0
      3s   1851.8        0       30      10/0       8/0       8/0       4/0
      4s   1988.7        0       42      13/0      11/0      10/0       8/0
      5s   1742.7        0       42      13/0      11/0      10/0       8/0
      6s   1881.6        0       68      18/0      17/0      17/0      16/0
      7s   1892.6        0       68      18/0      17/0      17/0      16/0
      8s   1926.2        0       67      17/0      17/0      17/0      16/0
      9s   1745.6        0       70      18/0      18/0      18/0      16/0
     10s   1895.7        0      102     25/15      27/7      26/6      24/2
     11s   1917.8        0      114     27/15      29/7      30/6      28/2
     12s   1874.6        0      114     27/15      29/7      30/6      28/2
     13s   1946.5        0      114     27/15      29/7      30/6      28/2
     14s   1872.8        0      114     27/15      29/7      30/6      28/2
     15s   1888.3        0      114     27/15      29/7      30/6      28/2
     16s   1879.1        0      114     27/15      29/7      30/6      28/2
     17s   1866.3        0      129     32/15      32/7      34/6      31/2
     18s   1801.5        0      159     39/15      41/7      41/6      38/2
     19s   1844.7        0      186     45/15      49/7      48/6      44/2
     20s   1775.8        0      204     48/22     54/16     53/14     49/11
_elapsed__ops/sec___errors_replicas_________0_________1_________2_________3
     21s   1860.9        0      210     50/22     54/16     55/14     51/11
     22s   1891.3        0      210     50/22     54/16     55/14     51/11
     23s   1945.9        0      210     50/22     54/16     55/14     51/11
     24s   1785.0        0      210     50/22     54/16     55/14     51/11
     25s   1815.1        0      210     50/22     54/16     55/14     51/11
     26s   1914.4        0      210     50/22     54/16     55/14     51/11
     27s   1737.8        0      210     50/22     54/16     55/14     51/11
     28s   1852.8        0      210     50/22     54/16     55/14     51/11
___stats____________________________________0_________1_________2_________3
replicas  (total% / diff%)           23.8/1.2  25.7/0.7  26.2/1.2  24.3/0.7
  leases  (total% / diff%)           34.9/9.9  25.4/0.4  22.2/2.8  17.5/7.5
```

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-:-KY_CU40FfrzCuC6Qm-J:b-2ctv3y)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/cmd/allocsim/main.go, line 308 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYV3v_6-li8mJrRE_sg:-KYemkfG-WZC_N84yyJS:b2qahn6) ([raw file](https://github.com/cockroachdb/cockroach/blob/99397ffd58c5fbc252d381cb2c572dfb6d2e3c94/pkg/cmd/allocsim/main.go#L308)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Hmmm... I'm sure I moved it down.  Must not have saved.
</blockquote></details>

fixed for reals this time.

---

*[pkg/cmd/allocsim/main.go, line 345 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYZeovitd4_Z6-1EwvJ:-KYemjpXl8ToCEn10Vux:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f455a177f0eb414fcea09fadc01919fc7db5277/pkg/cmd/allocsim/main.go#L345)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is exiting (1) appropriate after the duration expires?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/allocsim/main.go, line 194 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYkgR45nvmPsTq3g52N:-KYkgR46jJ5u1Ka1hhi4:bupdqmv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f069ad43743992f2c8bd40e2fb5b0c6ad60af44/pkg/cmd/allocsim/main.go#L194)):*
> ```Go
> 			fmt.Fprintf(&buf, " %8.8s", input)
> 		default:
> 			// All replica and lease counts need 9 characters instead of 8.
> ```

This comment is lacking context. I'd probably just remove it as it is likely to rot.

---

*[pkg/cmd/allocsim/main.go, line 238 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYkkK6Ri6sc3DAwN6K8:-KYkkK6Ri6sc3DAwN6K9:btmgtn5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f069ad43743992f2c8bd40e2fb5b0c6ad60af44/pkg/cmd/allocsim/main.go#L238)):*
> ```Go
> 				post = deadPost
> 			}
> 			row[4+i] = fmt.Sprintf("%s%d/%d%s", pre, replicas[i], leases[i], post)
> ```

Won't the escape sequences get messed up by the `%9.9s` formatting directive used to place this entry in the line? As I mention below, I'm not sure this change to the output generation was a good idea. Adding the output of the final stats could have been done separately. 

---

*[pkg/cmd/allocsim/main.go, line 276 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYkjtlUqwQYyYtYOZEi:-KYkjtlUqwQYyYtYOZEj:b-6167cg) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f069ad43743992f2c8bd40e2fb5b0c6ad60af44/pkg/cmd/allocsim/main.go#L276)):*
> ```Go
> 	defer a.ranges.Unlock()
> 
> 	// TOTO(bram): With the addition of localities, these stats will have to be
> ```

`TOTO`?

---

*[pkg/cmd/allocsim/main.go, line 302 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYkinhse5pC_PdHZ9Gs:-KYkinhse5pC_PdHZ9Gt:b13vs6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f069ad43743992f2c8bd40e2fb5b0c6ad60af44/pkg/cmd/allocsim/main.go#L302)):*
> ```Go
> 				fromMean = math.Abs((float64(count) - mean) / total * 100)
> 			}
> 			row[i+4] = fmt.Sprintf("%#2.1f/%#2.1f", percent, fromMean)
> ```

What does `#` do for the `%f` formatting verb? I couldn't find an answer in the go docs and I'm not familiar with that usage. I wonder if the 3 digits of precision is necessary here. 

---

*[pkg/cmd/allocsim/main.go, line 304 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYkjTpRfJ0K-4qqd40A:-KYkjTpSZsk56ptsTU1s:b2mamf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f069ad43743992f2c8bd40e2fb5b0c6ad60af44/pkg/cmd/allocsim/main.go#L304)):*
> ```Go
> 			row[i+4] = fmt.Sprintf("%#2.1f/%#2.1f", percent, fromMean)
> 		}
> 		printLine(row, false)
> ```

You could have done the formatting here without introducing the `printLine` method. I think the refactoring resulted in more code:

```
var buf bytes.Buffer
fmt.Fprintf(&buf, '%8s  (total% / diff%)         ', name)
for i, count := range counts {
  var percent, fromMean float64
  if total != 0 {
    percent = float64(count) / total * 100
    fromMean = math.Abs((float64(count) - mean) / total * 100)
  }
  fmt.Fprintf(&buf, " %9.9s", fmt.Sprintf("%2.1f/%2.1f", percent, fromMean))
}
fmt.Println(buf.String())
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/cmd/allocsim/main.go, line 194 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYkgR45nvmPsTq3g52N:-KYlQxLoY4FYRjneIJ7g:b8o7888) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f069ad43743992f2c8bd40e2fb5b0c6ad60af44/pkg/cmd/allocsim/main.go#L194)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This comment is lacking context. I'd probably just remove it as it is likely to rot.
</blockquote></details>

just dumping the function

---

*[pkg/cmd/allocsim/main.go, line 238 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYkkK6Ri6sc3DAwN6K8:-KYlQjUA5VXqBP_0C52J:bw7mq1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f069ad43743992f2c8bd40e2fb5b0c6ad60af44/pkg/cmd/allocsim/main.go#L238)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Won't the escape sequences get messed up by the `%9.9s` formatting directive used to place this entry in the line? As I mention below, I'm not sure this change to the output generation was a good idea. Adding the output of the final stats could have been done separately. 
</blockquote></details>

You're correct, the escape sequence ruins the formatting directive.  I'm just scrapping all this.  

---

*[pkg/cmd/allocsim/main.go, line 276 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYkjtlUqwQYyYtYOZEi:-KYlPZhYaFugVYyM4VVe:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f069ad43743992f2c8bd40e2fb5b0c6ad60af44/pkg/cmd/allocsim/main.go#L276)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`TOTO`?
</blockquote></details>

Fixed.

---

*[pkg/cmd/allocsim/main.go, line 302 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYkinhse5pC_PdHZ9Gs:-KYlFCxXGyKLCwf31exC:b1hqeux) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f069ad43743992f2c8bd40e2fb5b0c6ad60af44/pkg/cmd/allocsim/main.go#L302)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

What does `#` do for the `%f` formatting verb? I couldn't find an answer in the go docs and I'm not familiar with that usage. I wonder if the 3 digits of precision is necessary here. 
</blockquote></details>

That was an experiment that I thought I had removed. 

Also, I got rid of the 3 digits of precision.  I can round them if you prefer instead of the implicit floor.

---

*[pkg/cmd/allocsim/main.go, line 304 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYkjTpRfJ0K-4qqd40A:-KYlPBsRFPiz2yRP8er1:b-nnoare) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f069ad43743992f2c8bd40e2fb5b0c6ad60af44/pkg/cmd/allocsim/main.go#L304)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You could have done the formatting here without introducing the `printLine` method. I think the refactoring resulted in more code:

```
var buf bytes.Buffer
fmt.Fprintf(&buf, '%8s  (total% / diff%)         ', name)
for i, count := range counts {
  var percent, fromMean float64
  if total != 0 {
    percent = float64(count) / total * 100
    fromMean = math.Abs((float64(count) - mean) / total * 100)
  }
  fmt.Fprintf(&buf, " %9.9s", fmt.Sprintf("%2.1f/%2.1f", percent, fromMean))
}
fmt.Println(buf.String())
```
</blockquote></details>

Sure.  I don't find it as elegant, but you are correct, a lot less code.  I've removed the printLine method and restored the original printing back.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179)*
<!-- Sent from Reviewable.io -->

petermattis commented on pkg/cmd/allocsim/main.go:
```diff
@@ -243,6 +241,37 @@ func (a *allocSim) monitor(d time.Duration) {
+	// TODO(bram): With the addition of localities, these stats will have to be
+	// updated.
+
+	fmt.Println(formatHeader("___stats___________________________", *numNodes))
```
`s/*numNodes/len(s.ranges.replicas)/g`

petermattis reviewed (approved):
LGTM

BramGruneir commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/allocsim/main.go, line 251 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-KYUJGfMKh9A0aybafBM-r6-251:-KYno4y6KxZHiiuucoKv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c96f515b6212b60e9da07988ead665b211e6c082/pkg/cmd/allocsim/main.go#L251)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/*numNodes/len(s.ranges.replicas)/g`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12179#-:-KYnqZVWGkM_W7r_UqjX:bnfp4nl)*
<!-- Sent from Reviewable.io -->
